### ISSUE TITLE
Why my restoring does not work as expected? 

### ISSUE THREAD

**Issue by simonfqy** [deepchem/deepchem#1094]
To implement early stopping, I have expanded the `fit()` function in `deepchem/models/tensorgraph/tensor_graph.py` in my customized implementation. The original code can be found here: https://github.com/deepchem/deepchem/blob/6f86377d9bff4541ff50ec014743d840a3e230f4/deepchem/models/tensorgraph/tensor_graph.py#L112

Here is a very simplified version of what I wrote inside the body of the function (this code cannot run, just present you with the most relevant information):
```python
if early_stopping:
    # Creating this new field called temp_model_dir 
    self.temp_model_dir = self.model_dir + "/temp"
       
    with self._get_tf("Graph").as_default():
      # The line below is simply added to prevent the error of "nothing to save"
      v1 = tf.get_variable("v1", shape=[3], initializer = tf.zeros_initializer)
      saver = tf.train.Saver()
     
      for _ in range(iterations):              
        # Temporarily overriding the path of saving.
        self.save_file = "%s/%s" % (self.temp_model_dir, "model")
        self.fit_generator(self.default_generator())
        # Validation score.
        val_sc = self.evaluate(valid_dataset, metric, transformers)[metric[0].name] 

        # Resuming the path of saving.
        self.save_file = "%s/%s" % (self.model_dir, "model")
                       
        if val_sc beats the current best score:
          best_score = val_sc
          # Saving in the resumed path
          saver.save(self.session, self.save_file, global_step=self.global_step)
       
      self.restore() # From source code it seems to only restore from the self.model_dir
      self.evaluate(valid_dataset, metric, transformers)
    return
```

As is shown, I have overriden the saving path (`self.save_file`) with `self.temp_model_dir` when running the `fit_generator()`, and only saving the session to `self.model_dir` after confirming that it is indeed the best model. When running the model, I also confirmed that it has the desired behavior: only the models producing the best scores (at the time) are saved in (the directory corresponding to) `self.model_dir`, while `self.temp_model_dir` has the whole collection of models saved automatically by `fit_generator()` function.

The `self.restore()` function which is defined in the same script restores the session from `self.model_dir` by default. So I would expect that only the model with the optimal number of training epochs should be restored by the `self.restore()`. However, by examining the results of the ensuing `self.evaluate()` function, it becomes clear that this is not the case: whether or not the last training epoch yields the best model, the `self.evaluate()` function in the second last line always produces a result identical to the last run of `self.evaluate(valid_dataset, metric, transformers)`.   

If any of you needs a runnable version of the code, please don't hesitate to ask me.

I really need your help in this. Thank you very much!

**Comment by simonfqy** [comment]
@lilleswing @rbharath Anyone of you can help me figure it out? I have further simplified the code above. Thanks!

**Comment by rbharath** [comment]
@simonfqy I believe you're trying to implement early stopping, but are running into issues loading the correct checkpoint?

Could you put together a minimal failing example? In general, it's hard to read a large mass of code and debug the issue.

**Comment by simonfqy** [comment]
@rbharath I have just further simplified it. Perhaps this is the best I can do, further simplifications may result in losing some important information.

**Comment by simonfqy** [comment]
I think I have found out the source of the problem. The code below would behave as I desired:
```python
if early_stopping:
    # Creating this new field called temp_model_dir 
    self.temp_model_dir = self.model_dir + "/temp"
       
    with self._get_tf("Graph").as_default():  
           
      for _ in range(iterations):              
        # Temporarily overriding the path of saving.
        self.save_file = "%s/%s" % (self.temp_model_dir, "model")
        self.fit_generator(self.default_generator())
        saver = tf.train.Saver()
        # Validation score.
        val_sc = self.evaluate(valid_dataset, metric, transformers)[metric[0].name] 

        # Resuming the path of saving.
        self.save_file = "%s/%s" % (self.model_dir, "model")
                       
        if val_sc beats the current best score:
          best_score = val_sc
          # Saving in the resumed path
          saver.save(self.session, self.save_file, global_step=self.global_step)
       
      self.restore() # From source code it seems to only restore from the self.model_dir
      self.evaluate(valid_dataset, metric, transformers)
    return
```
The reason my previous code does not work is that I have initialized the `saver` before the `fit_generator()`, at which point the graph only has 1 global variable, which is `v1`, and the `saver` is actually saving nothing else but the value of `v1` each time `saver.save()` is called.
